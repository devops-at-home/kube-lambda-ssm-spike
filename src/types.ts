import { OidcConfig } from './stacks/oidc.stack';

export type AppFactoryProps = {
    oidcConfig?: OidcConfig;
    environment: Environment;
    account: string;
    region: string;
};

const environments = ['test', 'prod'] as const;
export type Environment = (typeof environments)[number];
