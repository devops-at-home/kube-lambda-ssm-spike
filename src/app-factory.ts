import { App, StackProps } from 'aws-cdk-lib';
import { DockerImageStack } from './stacks/docker-image-stack';
// import { LambdaSSMStack } from './stacks/lambda-stack';
import { OIDCStack } from './stacks/oidc.stack';
import { AppFactoryProps } from './types';

const stackPrefix = 'KubeLambdaSSM';

export const appFactory = (app: App, props: AppFactoryProps) => {
    const { environment, account, region } = props;

    const stackProps: StackProps = {
        env: { account, region },
        tags: {
            environment,
        },
    };

    if (props.oidcConfig) {
        new OIDCStack(app, `${stackPrefix}-OIDCStack`, {
            ...stackProps,
            ...props.oidcConfig,
        });
    }

    // const { repository } =
    new DockerImageStack(app, `${stackPrefix}-DockerImageStack`, {});

    // new LambdaSSMStack(app, `${stackPrefix}-LambdaSSMStack`, {
    //     repository,
    // });
};
