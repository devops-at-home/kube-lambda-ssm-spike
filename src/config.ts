import { OidcConfig } from './stacks/oidc.stack';
import { AppFactoryProps } from './types';

const oidcConfig: OidcConfig = {
    owner: 'devops-at-home',
    repo: 'kube-lambda-ssm-spike',
    // filter: 'ref:/refs/head/main', TODO: tighten permissions here
    filter: '*',
    lookup: true,
};

const account = process.env.CDK_DEFAULT_ACCOUNT!;
const region = 'ap-southeast-2';

export const config = (): AppFactoryProps[] => {
    return [
        // test
        {
            account,
            region,
            environment: 'test',
            oidcConfig,
        },
    ];
};
