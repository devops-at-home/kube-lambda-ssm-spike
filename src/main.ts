import { App } from 'aws-cdk-lib';
import { appFactory } from './app-factory';
import { config } from './config';

const app = new App();

const appFactoryProps = config();

appFactoryProps.forEach((props) => {
    appFactory(app, props);
});

app.synth();
