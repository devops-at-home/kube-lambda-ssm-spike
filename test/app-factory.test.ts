import { App } from 'aws-cdk-lib';
import { appFactory } from '../src/app-factory';
import { config } from '../src/config';

describe('AppFactory snapshot tests', () => {
    const appFactoryProps = config();
    appFactoryProps.forEach((props) => {
        const app = new App();
        appFactory(app, props);
        const { stacks } = app.synth();
        stacks.forEach(({ template, stackName }) => {
            test(stackName, () => {
                expect(template).toMatchSnapshot();
            });
        });
    });
});
