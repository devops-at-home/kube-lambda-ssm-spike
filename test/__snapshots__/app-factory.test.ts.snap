// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppFactory snapshot tests KubeLambdaSSM-DockerImageStack 1`] = `
{
  "Outputs": {
    "ImageTag": {
      "Value": "11a7951562a4a5dad9067a6e1bc9191f304a5c571db091d9bd2422dbd5488060",
    },
    "RepositoryArn": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":ecr:ap-southeast-2:",
            {
              "Ref": "AWS::AccountId",
            },
            ":repository/",
            {
              "Fn::Sub": "cdk-hnb659fds-container-assets-\${AWS::AccountId}-ap-southeast-2",
            },
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`AppFactory snapshot tests KubeLambdaSSM-LambdaSSMStack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "KubeLambdaA65398E6": {
      "DependsOn": [
        "KubeLambdaServiceRole25DF37D5",
      ],
      "Properties": {
        "Architectures": [
          "x86_64",
        ],
        "Code": {
          "ImageUri": {
            "Fn::Join": [
              "",
              [
                {
                  "Ref": "AWS::AccountId",
                },
                ".dkr.ecr.ap-southeast-2.",
                {
                  "Ref": "AWS::URLSuffix",
                },
                "/",
                {
                  "Fn::Sub": "cdk-hnb659fds-container-assets-\${AWS::AccountId}-ap-southeast-2",
                },
                ":11a7951562a4a5dad9067a6e1bc9191f304a5c571db091d9bd2422dbd5488060",
              ],
            ],
          },
        },
        "Environment": {
          "Variables": {
            "KUBECONFIG": "/tmp/k3s.yaml",
          },
        },
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "KubeLambdaServiceRole25DF37D5",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "KubeLambdaLogRetention2EB15411": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "KubeLambdaA65398E6",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "KubeLambdaServiceRole25DF37D5": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaSSMPermissions5297CD06": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ssm:DescribeInstanceInformation",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "ssm:*Session",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "ssm:DescribeParameters",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "ssm:GetParameters",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/KUBECONFIG",
                  ],
                ],
              },
            },
            {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:kms:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":key/149bd7e6-b504-4688-b693-9dfbfe95bf37",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaSSMPermissions5297CD06",
        "Roles": [
          {
            "Ref": "KubeLambdaServiceRole25DF37D5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-southeast-2",
          },
          "S3Key": "5fa1330271b8967d9254ba2d4a07144f8acefe8b77e6d6bba38261373a50d5f8.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`AppFactory snapshot tests KubeLambdaSSM-OIDCStack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DeployRole885297C3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:devops-at-home/kube-lambda-ssm-spike:*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":oidc-provider/token.actions.githubusercontent.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
        "MaxSessionDuration": 7200,
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
